django/forms/formsets.py

```diff
@@ -105,6 +105,33 @@ def management_form(self):
             })
         return form
 
+    def _extract_form_indices(self):
+        """Find indices of all submitted forms"""
+        prefix = self.prefix + '-'
+        prefix_len = len(prefix)
+        for key in self.data.keys():
+            if key.startswith(prefix):
+                identifier = key[prefix_len:].split('-', 1)[0]
+                if identifier.isdigit():
+                    yield int(identifier)
+
+    def _extract_total_form_count(self):
+        """Extract the number of submitted forms"""
+        if self.management_form.cleaned_data[TOTAL_FORM_COUNT]:
+            # prefer the management form if data is present
+            return self.management_form.cleaned_data[TOTAL_FORM_COUNT]
+        try:
+            return max(self._extract_form_indices()) + 1
+        except ValueError:  # no forms submitted
+            return 0
+
+    def _extract_initial_form_count(self):
+        """Extract the number of initial forms"""
+        if self.management_form.cleaned_data[INITIAL_FORM_COUNT]:
+            # prefer the management form if data is present
+            return self.management_form.cleaned_data[INITIAL_FORM_COUNT]
+        return 0
```

Comment:
Sorry, it's hard to get empathy right over the internet. During the sprints I've talked to Russell, Marc, Loïc and Honza. It's clear that we need some solution for the initial form case and for the case where all form fields are checkboxes (in this case not checking anything will result in POST being empty). What is not yet clear is how to do it in a way that is both better and backwards compatible.

The pull request is mostly a playground for others to explore, there's a discussion thread on the mailing list where I hope to get an agreement of what the APIs should look like before finishing the implementation.